File: M1/azor-chatdog-golang/files/session_files.go
package files

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// Message represents a message in universal format
type Message struct {
	Role  string `json:"role"`
	Parts []Part `json:"parts"`
}

// Part represents a text part
type Part struct {
	Text string `json:"text"`
}

// SessionLogEntry represents a single entry in session log
type SessionLogEntry struct {
	Role      string `json:"role"`
	Timestamp string `json:"timestamp"` // Keep as string for compatibility with Python format
	Text      string `json:"text"`
}

// SessionLogData represents the complete session log file
type SessionLogData struct {
	SessionID  string            `json:"session_id"`
	Model      string            `json:"model"`
	SystemRole string            `json:"system_role"`
	History    []SessionLogEntry `json:"history"`
}

// SessionInfo represents metadata about a session
type SessionInfo struct {
	ID            string
	MessagesCount int
	LastActivity  string
	Error         string
}

// LoadSessionHistory loads session history from JSON file
func LoadSessionHistory(sessionID string) ([]Message, error) {
	logFilename := filepath.Join(LogDir, fmt.Sprintf("%s-log.json", sessionID))

	if _, err := os.Stat(logFilename); os.IsNotExist(err) {
		return []Message{}, fmt.Errorf("session log file '%s' does not exist. Starting new session", logFilename)
	}

	data, err := os.ReadFile(logFilename)
	if err != nil {
		return []Message{}, fmt.Errorf("cannot read log file '%s': %w", logFilename, err)
	}

	var logData SessionLogData
	if err := json.Unmarshal(data, &logData); err != nil {
		return []Message{}, fmt.Errorf("cannot decode log file '%s': %w", logFilename, err)
	}

	// Convert to universal format
	var history []Message
	for _, entry := range logData.History {
		history = append(history, Message{
			Role:  entry.Role,
			Parts: []Part{{Text: entry.Text}},
		})
	}

	return history, nil
}

// SaveSessionHistory saves session history to JSON file
func SaveSessionHistory(sessionID string, history []Message, systemPrompt, modelName string) error {
	if len(history) < 2 {
		// Don't save empty/incomplete sessions
		return nil
	}

	logFilename := filepath.Join(LogDir, fmt.Sprintf("%s-log.json", sessionID))

	// Convert to session log format
	var jsonHistory []SessionLogEntry
	for _, msg := range history {
		if len(msg.Parts) > 0 {
			jsonHistory = append(jsonHistory, SessionLogEntry{
				Role:      msg.Role,
				Timestamp: time.Now().Format("2006-01-02T15:04:05.000000"),
				Text:      msg.Parts[0].Text,
			})
		}
	}

	logData := SessionLogData{
		SessionID:  sessionID,
		Model:      modelName,
		SystemRole: systemPrompt,
		History:    jsonHistory,
	}

	data, err := json.MarshalIndent(logData, "", "    ")
	if err != nil {
		return fmt.Errorf("error marshaling session data: %w", err)
	}

	if err := os.WriteFile(logFilename, data, 0644); err != nil {
		return fmt.Errorf("error writing to file %s: %w", logFilename, err)
	}

	return nil
}

// ListSessions returns a list of available sessions with metadata
func ListSessions() []SessionInfo {
	files, err := os.ReadDir(LogDir)
	if err != nil {
		return []SessionInfo{}
	}

	var sessionIDs []string
	for _, f := range files {
		if strings.HasSuffix(f.Name(), "-log.json") && f.Name() != "azor-wal.json" {
			sessionID := strings.TrimSuffix(f.Name(), "-log.json")
			sessionIDs = append(sessionIDs, sessionID)
		}
	}
	sort.Strings(sessionIDs)

	var sessions []SessionInfo
	for _, sid := range sessionIDs {
		logPath := filepath.Join(LogDir, fmt.Sprintf("%s-log.json", sid))

		data, err := os.ReadFile(logPath)
		if err != nil {
			sessions = append(sessions, SessionInfo{
				ID:    sid,
				Error: "BŁĄD ODCZYTU PLIKU",
			})
			continue
		}

		var logData SessionLogData
		if err := json.Unmarshal(data, &logData); err != nil {
			sessions = append(sessions, SessionInfo{
				ID:    sid,
				Error: "BŁĄD ODCZYTU PLIKU",
			})
			continue
		}

		timeStr := "Brak aktywności"
		if len(logData.History) > 0 {
			lastMsg := logData.History[len(logData.History)-1]
			// Parse timestamp string
			if lastMsg.Timestamp != "" {
				// Try parsing Python ISO format
				t, err := time.Parse("2006-01-02T15:04:05.999999", lastMsg.Timestamp)
				if err == nil {
					timeStr = t.Format("2006-01-02 15:04")
				} else {
					// Try with timezone
					t, err = time.Parse(time.RFC3339, lastMsg.Timestamp)
					if err == nil {
						timeStr = t.Format("2006-01-02 15:04")
					}
				}
			}
		}

		sessions = append(sessions, SessionInfo{
			ID:            sid,
			MessagesCount: len(logData.History),
			LastActivity:  timeStr,
		})
	}

	return sessions
}

// RemoveSessionFile removes a session log file
func RemoveSessionFile(sessionID string) error {
	logFilename := filepath.Join(LogDir, fmt.Sprintf("%s-log.json", sessionID))

	if _, err := os.Stat(logFilename); os.IsNotExist(err) {
		return fmt.Errorf("session file for ID '%s' not found", sessionID)
	}

	if err := os.Remove(logFilename); err != nil {
		return fmt.Errorf("error removing session file '%s': %w", logFilename, err)
	}

	return nil
}

File: M1/azor-chatdog-py/src/files/session_files.py
import os
import json
from dataclasses import dataclass
from datetime import datetime
from typing import List, Any, Dict
from files.config import LOG_DIR
from assistant import DEFAULT_ASSISTANT_NAME


@dataclass
class SessionData:
    history: List[Dict]
    title: str | None
    assistant_name: str
    system_role: str | None
    error: str | None


def load_session(session_id: str) -> SessionData:
    """Loads session data from a JSON file in universal format."""
    log_filename = os.path.join(LOG_DIR, f"{session_id}-log.json")
    if not os.path.exists(log_filename):
        return SessionData(history=[], title=None, assistant_name=DEFAULT_ASSISTANT_NAME, system_role=None, error=f"Session log file '{log_filename}' does not exist. Starting new session.")

    try:
        with open(log_filename, 'r', encoding='utf-8') as f:
            log_data = json.load(f)
    except json.JSONDecodeError:
        return SessionData(history=[], title=None, assistant_name=DEFAULT_ASSISTANT_NAME, system_role=None, error=f"Cannot decode log file '{log_filename}'. Starting new session.")

    history = []
    for entry in log_data.get('history', []):
        history.append({
            "role": entry['role'],
            "parts": [{"text": entry['text']}]
        })

    return SessionData(
        history=history,
        title=log_data.get('title'),
        assistant_name=log_data.get('assistant_name', DEFAULT_ASSISTANT_NAME),
        system_role=log_data.get('system_role'),
        error=None,
    )

def save_session_history(session_id: str, history: List[Dict], system_prompt: str, model_name: str, assistant_name: str, title: str | None = None) -> tuple[bool, str | None]:
    """
    Saves the current session history to a JSON file,
    only if the history contains at least one complete turn (User + Model).
    
    Args:
        session_id: Unique identifier for the session
        history: Conversation history to save (universal format: List of dicts)
        system_prompt: System prompt used for the assistant
        model_name: Name of the LLM model used
    
    Returns:
        tuple[bool, str | None]: (success, error_message)
    """
    if len(history) < 2:
        # CONDITION: Prevents saving empty/incomplete session
        return True, None

    log_filename = os.path.join(LOG_DIR, f"{session_id}-log.json")

    json_history = []
    for content in history:
        # Handle universal format (dictionaries) from both Gemini and LLaMA clients
        if isinstance(content, dict) and 'parts' in content and content['parts']:
            text_part = content['parts'][0].get('text', '')
        else:
            # Fallback for legacy formats
            text_part = getattr(content, 'parts', [{}])[0].get('text', '') if hasattr(content, 'parts') else ""
        
        if text_part:
            json_history.append({
                'role': content.get('role', '') if isinstance(content, dict) else getattr(content, 'role', ''),
                'timestamp': datetime.now().isoformat(),
                'text': text_part
            })

    log_data = {
        'session_id': session_id,
        'title': title,
        'assistant_name': assistant_name,
        'model': model_name,
        'system_role': system_prompt,
        'history': json_history
    }

    try:
        with open(log_filename, 'w', encoding='utf-8') as f:
            json.dump(log_data, f, indent=4, ensure_ascii=False)
        return True, None
        
    except Exception as e:
        return False, f"Error writing to file {log_filename}: {e}"

def list_sessions():
    """Returns a list of available sessions with their metadata."""
    files = os.listdir(LOG_DIR)
    session_ids = sorted([f.replace('-log.json', '') for f in files if f.endswith('-log.json') and f != 'azor-wal.json'])
    
    sessions_data = []
    for sid in session_ids:
        log_path = os.path.join(LOG_DIR, f"{sid}-log.json")
        try:
            with open(log_path, 'r', encoding='utf-8') as f:
                log_data = json.load(f)
                history_len = len(log_data.get('history', []))
                last_msg_time_str = log_data.get('history', [{}])[-1].get('timestamp', 'Brak daty')
                
                time_str = 'Brak aktywności'
                if last_msg_time_str != 'Brak daty':
                    try:
                        dt = datetime.fromisoformat(last_msg_time_str)
                        time_str = dt.strftime('%Y-%m-%d %H:%M')
                    except ValueError:
                        pass
            
            sessions_data.append({
                'id': sid,
                'title': log_data.get('title'),
                'messages_count': history_len,
                'last_activity': time_str,
                'error': None
            })
        except Exception:
            sessions_data.append({
                'id': sid,
                'error': 'BŁĄD ODCZYTU PLIKU'
            })
    
    return sessions_data

def remove_session_file(session_id: str) -> tuple[bool, str | None]:
    """
    Removes a session log file from the filesystem.

    Args:
        session_id: The ID of the session to remove.

    Returns:
        A tuple containing a boolean indicating success and an optional error message.
    """
    log_filename = os.path.join(LOG_DIR, f"{session_id}-log.json")
    if not os.path.exists(log_filename):
        return False, f"Session file for ID '{session_id}' not found."

    try:
        os.remove(log_filename)
        return True, None
    except OSError as e:
        return False, f"Error removing session file '{log_filename}': {e}"
